sklearn/utils/__init__.py

```diff
@@ -178,7 +178,42 @@ def axis0_safe_slice(X, mask, len_mask):
     return np.zeros(shape=(0, X.shape[1]))
 
 
-def safe_indexing(X, indices):
+def safe_indexing(X, indices, axis=0):
+    """Return rows, items or columns of X using indices.
+
+    Parameters
+    ----------
+    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series.
+        Data from which to sample rows, items or columns.
+    indices : see below
+        When ``axis=0``, indices need to be an array of integer.
+        When ``axis=1``, indices can be one of:
+            Supported key types (key):
+            - scalar: output is 1D
+            - lists, slices, boolean masks: output is 2D
+            - callable that returns any of the above
+            Supported key data types:
+            - integer or boolean mask (positional):
+                - supported for arrays, sparse matrices and dataframes
+            - string (key-based):
+                - only supported for dataframes
+                - So no keys other than strings are allowed (while in principle
+                  you can use any hashable object as key).
+    axis : int, default=0
+        The axis along which the X will be subsampled. ``axis=0`` will select
+        rows while ``axis=1`` will select columns.
+
+    Notes
+    -----
+    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
+    not supported.
+    """
+    if axis == 0:
```

Comment:
> We could go pandas style and also support 'rows' and 'columns', it's always easier to read to me.

IMO, I also think it makes it easier to read. However, I would not introduce it here because we will need to be consistent with the other utils as well. Supporting the numpy style with good documentation would be already a plus.

Reply:
I don't immediately see other utils supporting an `axis` parameter?
Also I don't think consistency is such an issue here. That would be an incentive for the other potential utils to implement this too.

Reply:
You have `weighted_mode`, `incr_mean_variance_axis`, `mean_variance_axis` for the moment.
However, I have a second thought. It makes sense to not have support for `columns` and `rows` for these ones since you cannot input dataframe and we could only add support on helper supporting dataframe.

I would still let this for another PR since it would add some code for checking that we are not using the different arguments in conjunction, making the diff quite larger.

Reply:
I don't see how this would only be useful for dataframes?
and I think the change would be as easy as 

```py
if axis == 'rows':
    axis = 0
elif axis == 'columns':
	axis = 1
```

but OK, won't push it more.